File: lab_8/zadanie3.java
class Pizza {

    private String name;
    private String dough;
    private String sauce;
    private String topping;

    private Pizza(PizzaBuilder builder) {
        this.name = builder.name;
        this.dough = builder.dough;
        this.sauce = builder.sauce;
        this.topping = builder.topping;
    }

    @Override
    public String toString() {
        return "Pizza [name=" + name + ", dough=" + dough + ", sauce=" + sauce + ", topping=" + topping + "]";
    }

    public static class PizzaBuilder {

        private String name;
        private String dough;
        private String sauce;
        private String topping;

        public PizzaBuilder name(String name) {
            this.name = name;
            return this;
        }

        public PizzaBuilder dough(String dough) {
            this.dough = dough;
            return this;
        }

        public PizzaBuilder sauce(String sauce) {
            this.sauce = sauce;
            return this;
        }

        public PizzaBuilder topping(String topping) {
            this.topping = topping;
            return this;
        }

        public Pizza build() {
            return new Pizza(this);
        }
    }
}

public class Main {
    public static void main(String[] args) {
        Pizza pepperoniPizza = new Pizza.PizzaBuilder()
                .name("Pepperoni Pizza")
                .dough("Thin crust")
                .sauce("Tomato")
                .topping("Pepperoni")
                .build();

        Pizza margaritaPizza = new Pizza.PizzaBuilder()
                .name("Margarita Pizza")
                .dough("Regular crust")
                .sauce("Tomato")
                .topping("Cheese")
                .build();

        System.out.println(pepperoniPizza);
        System.out.println(margaritaPizza);
    }
}

File: lab_8/zadanie3.cpp
#include <iostream>
#include <string>

class Pizza {
private:
    std::string name;
    std::string dough;
    std::string sauce;
    std::string topping;

    Pizza(const std::string& name, const std::string& dough, const std::string& sauce, const std::string& topping)
        : name(name), dough(dough), sauce(sauce), topping(topping) {}

public:
    class Builder { // ten wzorzec Budowniczego powinien byc ukonkretniony dlam różnych rodzajów pizzy
    private:
        std::string name;
        std::string dough;
        std::string sauce;
        std::string topping;

    public:
        Builder& setName(const std::string& n) {
            name = n;
            return *this;
        }

        Builder& setDough(const std::string& d) {
            dough = d;
            return *this;
        }

        Builder& setSauce(const std::string& s) {
            sauce = s;
            return *this;
        }

        Builder& setTopping(const std::string& t) {
            topping = t;
            return *this;
        }

        Pizza build() {
            return Pizza(name, dough, sauce, topping);
        }
    };

    void display() const {
        std::cout << name << " Pizza with " 
                  << dough << " dough, " 
                  << sauce << " sauce, and " 
                  << topping << " topping." << std::endl;
    }
};

int main() {

    Pizza margarita = Pizza::Builder()
                        .setName("Margarita")
                        .setDough("Thin crust")
                        .setSauce("Tomato")
                        .setTopping("Cheese")
                        .build(); // ???

    margarita.display();

    Pizza pepperoni = Pizza::Builder()
                        .setName("Pepperoni")
                        .setDough("Thick crust")
                        .setSauce("Barbecue")
                        .setTopping("Pepperoni")
                        .build();  // ???

    pepperoni.display();

    return 0;
}
